### PR TITLE
b_thread deepcopy bug fix and external event support

### PR DESCRIPTION
**Pull request by tomyaacov**

### CODE REVIEW

**Comment by tomyaacov** [comment]
Hi @adielashrov. please review this pull request. it contains a fix for the thread deepcopy issue and external events support. We still need to figure out how external event support acts in rich events protocol.
In any case, let me know what you think (don't merge the pull request yet)

File: bppy/model/event_selection/simple_event_selection_strategy.py [hunk 1]
```diff
@@ -37,10 +37,13 @@ def selectable_events(self, statements):
...
         return possible_events
 
-    def select(self, statements):
+    def select(self, statements, external_events_queue=[]):
         selectable_events = self.selectable_events(statements)
         if selectable_events:
             return random.choice(tuple(selectable_events))
         else:
-            return None
+            if len(external_events_queue) > 0:
```
Comment (adielashrov) [inline bppy/model/event_selection/simple_event_selection_strategy.py:45]: This is the correct order of handling external events and internal events(checked with Guy).
1. finish the superstep 
2. Inject the next external event.
Nice Job!

File: bppy/model/event_selection/simple_event_selection_strategy.py [hunk 2]
```diff
@@ -37,10 +37,13 @@ def selectable_events(self, statements):
                     possible_events = {x for x in possible_events if x not in statement.get('block')}
         return possible_events
 
-    def select(self, statements):
+    def select(self, statements, external_events_queue=[]):
         selectable_events = self.selectable_events(statements)
         if selectable_events:
             return random.choice(tuple(selectable_events))
```
Comment (adielashrov) [inline bppy/model/event_selection/simple_event_selection_strategy.py:43]: Something I noticed.
If there are no possible selected events, random.choice will raise an exepction.
Example:
`import random
selectable_events = set()
print(type(selectable_events))
event = random.choice(tuple(selectable_events))`

Comment (adielashrov) [reply bppy/model/event_selection/simple_event_selection_strategy.py:43]:
```suggestion
        if len(possible_events) > 0:
            return possible_events
        else:
            return None
```

File: bppy/model/event_selection/simple_event_selection_strategy.py [hunk 3]
```diff
@@ -37,10 +37,13 @@ def selectable_events(self, statements):
                     possible_events = {x for x in possible_events if x not in statement.get('block')}
         return possible_events
```
Comment (adielashrov) [inline bppy/model/event_selection/simple_event_selection_strategy.py:38]:
```suggestion
        if len(possible_events) > 0:
            return possible_events
        else:
            return None
```

**Review by adielashrov — COMMENTED** [review]
Hey,
Added comments + a small fix to selectable_events method

File: bppy/model/event_selection/simple_event_selection_strategy.py (hunk 3 above)
Comment (adielashrov) [reply bppy/model/event_selection/simple_event_selection_strategy.py:38]: https://github.com/bThink-BGU/BPpy/issues/4

File: bppy/model/event_selection/simple_event_selection_strategy.py (hunk 2 above)
Comment (tomyaacov) [reply bppy/model/event_selection/simple_event_selection_strategy.py:43]: @adielashrov yeah. that's why we check if it is empty in the select method.
keep in mind that:
`
if []:
    print("has events")
else:
    print("no events")
`
return "no events"

File: bppy/model/event_selection/simple_event_selection_strategy.py (hunk 3 above)
Comment (tomyaacov) [reply bppy/model/event_selection/simple_event_selection_strategy.py:38]: **this check is done in the select method already:**
```python
    def select(self, statements, external_events_queue=[]):
        selectable_events = self.selectable_events(statements)
        if selectable_events:
            return random.choice(tuple(selectable_events))
        else:
            if len(external_events_queue) > 0:
                return external_events_queue.pop(0)
            else:
                return None
```

**Pull request merged**